python/ray/data/dataset.py

```diff
@@ -1759,9 +1759,40 @@ def limit(self, limit: int) -> "Dataset[T]":
         Returns:
             The truncated dataset.
         """
-
-        left, _ = self._split(limit, return_right_half=False)
-        return left
+        start_time = time.perf_counter()
+        # Truncate the block list to the minimum number of blocks that contains at least
+        # `limit` rows.
+        block_list = self._plan.execute().truncate_by_rows(limit)
```

Comment:
This should not execute at here at the moment.

Reply:
Are you suggesting that we turn the `.limit()` API into a (possibly lazy) transformation stage rather than an always-eager operation?

Note that executing the plan won't trigger reading.

Reply:
Right, if we run execute() regardless, this is still a consumption. But sounds fair to keep this out of scope for this PR.